### PR TITLE
reset the `__eq__` and `__hash__` of HTTPConnection to allow WebSockets to be added to …

### PR DESCRIPTION
**Pull request by graingert**
…sets

https://github.com/tiangolo/fastapi/pull/1991/files#diff-a3e1b3437a67897cf7539e9ec348c9d8R50

### CODE REVIEW

**Comment by graingert** [comment]
@JayH5 is this a good change?

**Comment by lovelydinosaur** [comment]
Could you explain the motivation a bit more clearly here?

**Comment by graingert** [comment]
@tomchristie for a number of reasons it's really useful to be able to put connections in a set. Especially websocket pubsub, see https://github.com/tiangolo/fastapi/pull/1991

I can see no use for comparing non-identical connections for equality

File: starlette/requests.py [hunk 1]
```diff
@@ -73,6 +73,9 @@ def __iter__(self) -> typing.Iterator[str]:
     def __len__(self) -> int:
         return len(self.scope)
 
+    __eq__ = object.__eq__
+    __hash__ = object.__hash__
```
[the remark this answers is not in the document]
Comment (graingert) [reply starlette/requests.py:81]: Mapping overrides `__eq__` which disables the default `__hash__`

[the remark this answers is not in the document]
Comment (lovelydinosaur) [reply starlette/requests.py:81]: > Mapping overrides __eq__

Ah gotcha, thanks.
Does that mean we actually only need to override `__eq__` then?

What do we think to something like the following?...

```python
    # Don't use the `abc.Mapping.__eq__` implementation.
    # Connection instances should never be considered equal
    # unless `self is other`.
    __eq__ = object.__eq__
```

[the remark this answers is not in the document]
Comment (graingert) [reply starlette/requests.py:81]: I think you need to re-enable `__hash__` but I'll double check

[the remark this answers is not in the document]
Comment (graingert) [reply starlette/requests.py:81]: @tomchristie yep you need both:

```python
from __future__ import annotations

import collections.abc
import pytest

def test_default():
    class HTTPConnection(collections.abc.Mapping):
        def __getitem__(self, *args, **kwargs):
            return None

        def __iter__(self, *args, **kwargs):
            return self

        def __next__(self, *args, **kwargs):
            raise StopIteration

        def __len__(self, *args, **kwargs):
            return 0


    assert HTTPConnection() == HTTPConnection()
    with pytest.raises(TypeError, match=r"unhashable type: 'HTTPConnection'"):
        connections = {HTTPConnection()}


def test_eq():
    class HTTPConnection(collections.abc.Mapping):

        __eq__ = object.__eq__

        def __getitem__(self, *args, **kwargs):
            return None

        def __iter__(self, *args, **kwargs):
            return self

        def __next__(self, *args, **kwargs):
            raise StopIteration

        def __len__(self, *args, **kwargs):
            return 0


    assert HTTPConnection() != HTTPConnection()
    h = HTTPConnection()
    assert h == h

    with pytest.raises(TypeError, match=r"unhashable type: 'HTTPConnection'"):
        connections = {HTTPConnection()}



def test_eq_and_hash():
    class HTTPConnection(collections.abc.Mapping):

        __eq__ = object.__eq__
        __hash__ = object.__hash__

        def __getitem__(self, *args, **kwargs):
            return None

        def __iter__(self, *args, **kwargs):
            return self

        def __next__(self, *args, **kwargs):
            raise StopIteration

        def __len__(self, *args, **kwargs):
            return 0


    assert HTTPConnection() != HTTPConnection()
    h = HTTPConnection()
    assert h == h

    assert {HTTPConnection()} != {HTTPConnection()}
    assert h in {h}
    assert {h} == {h}
 ```

**Review by lovelydinosaur — APPROVED** [review]
Okay, fair enough. 😄

File: starlette/requests.py [hunk 2]
```diff
@@ -74,6 +74,9 @@ def __iter__(self) -> typing.Iterator[str]:
     def __len__(self) -> int:
         return len(self.scope)
 
+    __eq__ = object.__eq__
```
Comment (graingert) [inline starlette/requests.py:80]:
```suggestion
    # Don't use the `abc.Mapping.__eq__` implementation.
    # Connection instances should never be considered equal
    # unless `self is other`.
    __eq__ = object.__eq__
```

File: tests/test_websockets.py [hunk 3]
```diff
@@ -368,3 +368,13 @@ async def mock_send(message):
...
+
+    # check __eq__ and __hash__
+    assert websocket != WebSocket(
+        {"type": "websocket", "path": "/abc/", "headers": []},
+        receive=mock_receive,
+        send=mock_send,
+    )
+    assert websocket == websocket
+    assert websocket in {websocket}
+    assert {websocket} == {websocket}
```
Comment (lovelydinosaur) [inline tests/test_websockets.py:380]: Yup, that's all neat enough. 👍

**Review by lovelydinosaur — APPROVED** [review]

**Pull request merged**